src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,567 @@
+        movdqa          xmm3, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 30h]
+        
+        jmp             LOCAL_LABEL(HaveThread)
+        
+LOCAL_LABEL(DoTrapReturningThreadsTHROW):
+
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h], rdi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h], rsi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h], rdx
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 18h], rcx
+
+        // @CONSIDER: mark UMEntryThunks that have FP params and only save/restore xmm regs on those calls
+        //            initial measurements indidcate that this could be worth about a 5% savings in reverse
+        //            pinvoke overhead.
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h], xmm0
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 10h], xmm1
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 20h], xmm2
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 30h], xmm3
+
+        mov             [rbp + UMThunkStubAMD64_RARE_PATH_SPILL_OFFSET], METHODDESC_REGISTER
+        mov             rdi, r12                                                                  // Thread* pThread
+        mov             rsi, METHODDESC_REGISTER                                                  // UMEntryThunk* pUMEntry
+        call            C_FUNC(UMThunkStubRareDisableWorker)
+        mov             METHODDESC_REGISTER, [rbp + UMThunkStubAMD64_RARE_PATH_SPILL_OFFSET]
+
+        mov             rdi,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h] 
+        mov             rsi,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h] 
+        mov             rdx,   [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h] 
+        mov             rcx,   [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 18h]
+
+        // @CONSIDER: mark UMEntryThunks that have FP params and only save/restore xmm regs on those calls
+        movdqa          xmm0, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h]
+        movdqa          xmm1, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 10h]
+        movdqa          xmm2, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 20h]
+        movdqa          xmm3, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 30h]
+
+        jmp             LOCAL_LABEL(InCooperativeMode)
+
+LOCAL_LABEL(UMThunkStub_CopyStackArgs):
+        // rax = cbStackArgs (with 20h for register args subtracted out already)
+
+        sub             rsp, rax
+        and             rsp, -16
+
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h], rdi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h], rsi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h], rdx
```

Comment:
These are the registers that are used in CopyLoop. Please either save r8 instead of rdx or change CopyLoop to use rdx instead of r8.